### ISSUE TITLE
leanCoP import generates expansion sequents with empty term sets

### ISSUE THREAD

**Issue by gebner** [gapt/gapt#366]
@gisellemnr 
```scala
val p = LeanCoPParser.getExpansionProof("testing/TSTP/leanCoP/NUM/NUM333+1/leanCoP---2.2.THM-Prf.s.out").get
println(toShallow(p))
println(toDeep(p))
println(extractInstances(p)) // <- formulas in this sequent correspond to elements in the term set
```
```
 :- ∃_272634.∃_272631.∃_272628.∃_272625.∃_272622.∃_272619.∃_272616.((sum(_272616,_272619,_272625)∧(sum(_272625,_272622,_272628)∧(sum(_272619,_272622,_272631)∧sum(_272616,_272631,_272634))))⊃_272628=_272634)
 :- ⊥
 :- 
```

**Comment by gisellemnr** [comment]
This is a very odd proof actually... To make it easier to read, it is trying to prove that:

    :- ∃ x7. ∃ x6. ∃ x5. ∃ x4. ∃ x3. ∃ x2. ∃ x1
    sum(x1, x2, x4) ^ sum(x4, x3, x5) ^
    sum(x2, x3, x6) ^ sum(x1, x6, x7) 
    => x5 = x7

The proof leanCoP produces takes the equality axiom x=x and replaces x5 and x7 both by x. I guess this is not the proof the author of the theorem was after... It is strange, but it is a proof in any case, right? What do you think?

The problem with the expansion proof generated is that not all quantifiers are instantiated. So it generates a "partial" expansion proof, with nodes ending with quantified formulas, by instantiating only x7 (since x6 has no instance, it stops there...). What is the correct way to deal with this? Throw an exception? Warning? Use dummy constants? I guess we cannot build an LK proof or run cut-intro unless we have instances of all quantified variables, right?

@shetzl, can you take a look at this?

**Comment by gebner** [comment]
You're right, this is an odd theorem.

As far as I see, the reason why leancop doesn't instantiate the other variables is because it is actually proving a CNF of the original formula that contains x5=x7 as a unit clause.  We're however constructing an expansion proof of the original formula.

Regarding the expansion trees, we want the deep formula to be a quasi-tautology, hence I guess the only possible output in this proof is weak quantifier nodes with fresh constants/variables as dummy instances.

A similar situation can occur in the prover9 import: For example when proving the CNF P(x,y) & (P(x,y) -> Q(y)) & -Q(c), prover9 doesn't completely instantiate P(x,y) either; in the end the LK proof contains the instance P(x,c) with one free variable (and this variable also occurs in the extracted term set).

Exceptions or warnings are not a solution.  Inserting a weakening node instead of a weak quantifier node gives bottom as the deep formula, so this does not work either.

(As for cut-intro: it should be enough if the deep sequent is a tautology.)

**Comment by shetzl** [comment]
It is indeed an odd formula to prove. It every existential would be replaced by a universal (and sufficient background axioms would be supplied) it would express the associativity of addition in the formulation as ternary predicate (are we sure these should be existentials?)

As for partial instantiation: this is permitted in some formalisms (such as resolution) but it is not in expansion proofs. In my opinion it is the responsibility of the code that generates the expansion proof (i.e. the import in this case) to create a correct expansion proof - either by dummy constants or by identity substitutions. I rather lean towards dummy constants as it creates a ground deep formula.

Weakening in expansion proofs is something different: it designates that a part of the deep formula is not necessary for obtaining a tautology - this is not the point here: the part "x5 = x7" (properly instantiated) is needed. Unrelated to that, a weakening node could be used for the sum-conjunction.

**Comment by gisellemnr** [comment]
I checked the problem file, and it uses indeed existential quantifiers, although I think the idea is to use universal ones. I'll send a message to Geoff Sutcliffe about it.

Here is the input problem:

    %------------------------------------------------------------------------------
    % File     : NUM333+1 : TPTP v6.1.0. Released v3.1.0.
    % Domain   : Number Theory (RDN arithmetic)
    % Problem  : ?XYZ, ((X+Y)+Z) = (X+(Y+Z))
    % Version  : Especial.
    % English  :
    
    % Refs     :
    % Source   : [TPTP]
    % Names    :
    
    % Status   : Theorem
    % Rating   : 0.20 v6.1.0, 0.17 v6.0.0, 0.22 v5.5.0, 0.19 v5.4.0, 0.18 v5.3.0, 0.26 v5.2.0, 0.15 v5.1.0, 0.10 v5.0.0, 0.12 v4.1.0, 0.13 v4.0.0, 0.12 v3.7.0, 0.15 v3.5.0, 0.16 v3.4.0, 0.11 v3.3.0, 0.07 v3.2.0, 0.27 v3.1.0
    % Syntax   : Number of formulae    :  402 ( 374 unit)
    %            Number of atoms       :  477 (   6 equality)
    %            Maximal formula depth :   19 (   2 average)
    %            Number of connectives :   77 (   2 ~  ;   1  |;  46  &)
    %                                         (   3 <=>;  25 =>;   0 <=)
    %                                         (   0 <~>;   0 ~|;   0 ~&)
    %            Number of predicates  :   11 (   0 propositional; 1-4 arity)
    %            Number of functors    :  260 ( 256 constant; 0-2 arity)
    %            Number of variables   :  128 (   0 singleton; 121 !;   7 ?)
    %            Maximal term depth    :    5 (   2 average)
    % SPC      : FOF_THM_RFO_SEQ
    
    % Comments :
    %------------------------------------------------------------------------------
    %----Include axioms for RDN arithmetic
    include('Axioms/NUM005+0.ax').
    include('Axioms/NUM005+1.ax').
    include('Axioms/NUM005+2.ax').
    %------------------------------------------------------------------------------
    %Comment:S-SETHEO---0.0 gives error:
    %  *** ERROR:bigloo:parser:
    %  parse error (unexpected token `closebracket') -- closebracket
    fof(associative,conjecture,
        ( ? [X,Y,Z,Z1,Z2,Z3,Z4] :
            ( ( sum(X,Y,Z1)
              & sum(Z1,Z,Z2)
              & sum(Y,Z,Z3)
              & sum(X,Z3,Z4) )
           => Z2 = Z4 ) )).
    
    %------------------------------------------------------------------------------

I will think of how to complete these expansion trees to close this bug.

**Comment by gebner** [comment]
Ha, interestingly enough, the rewrite of `formulaToExpansionTree` fixed this issue.  The deep formula is now quasi-tautological and we can construct an LK proof from it.  This is the deep formula now:
```
scala> loadExpansionProof(file"testing".glob("**/NUM333+1/leanCoP*").next).deep
res16: at.logic.gapt.proofs.Sequent[at.logic.gapt.expr.HOLFormula] =

⊢
sum(#v(_272616: i), #v(_272619: i), #v(_272625: i)) ∧
    (sum(#v(_272625: i), #v(_272622: i), #v(_271429: i)) ∧
      (sum(#v(_272619: i), #v(_272622: i), #v(_272631: i)) ∧
        sum(#v(_272616: i), #v(_272631: i), #v(_271429: i)))) ⊃
  #v(_271429: i) = #v(_271429: i)
```
